lib/experimental/class-wp-navigation-fallbacks-gutenberg.php

```diff
@@ -0,0 +1,223 @@
+<?php
+/**
+ * WP_Navigation_Fallbacks_Gutenberg class
+ *
+ * @package gutenberg
+ * @since 6.3.0
+ */
+
+
+require __DIR__ . '/class-wp-menu-conversion-gutenberg.php';
+
+/**
+ * Manages Fallback behavior for Navigation menus.
+ *
+ * @access public
+ */
+class WP_Navigation_Fallbacks_Gutenberg {
+
+	/**
+	 * Gets (and/or creates) an appropriate fallback Navigation Menu.
+	 *
+	 * @return WP_Post|null the fallback Navigation Post or null.
+	 */
+	public static function get_fallback_menu() {
+
+		$fallback_sequence = array(
+			'get_most_recently_published_navigation',
+			'create_classic_menu_fallback',
+			'create_default_fallback',
+		);
+
+		// go through each of the fallbacks and return the first one that works.
+		foreach ( $fallback_sequence as $fallback ) {
+			$navigation_post = static::$fallback();
+
+			if ( $navigation_post && ! is_wp_error( $navigation_post ) ) {
+				return $navigation_post instanceof WP_Post ? $navigation_post : static::get_most_recently_published_navigation();
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * Finds the most recently published `wp_navigation` Post.
+	 *
+	 * @return WP_Post|null the first non-empty Navigation or null.
+	 */
+	public static function get_most_recently_published_navigation() {
+
+		// Default to the most recently created menu.
+		$parsed_args = array(
+			'post_type'              => 'wp_navigation',
+			'no_found_rows'          => true,
+			'update_post_meta_cache' => false,
+			'update_post_term_cache' => false,
+			'order'                  => 'DESC',
+			'orderby'                => 'date',
+			'post_status'            => 'publish',
+			'posts_per_page'         => 1, // get only the most recent.
+		);
+
+		$navigation_post = new WP_Query( $parsed_args );
+
+		if ( count( $navigation_post->posts ) > 0 ) {
+			return $navigation_post->posts[0];
+		}
+
+		return null;
+	}
+
+	/**
+	 * Creates a Navigation Menu post from a Classic Menu.
+	 *
+	 * @return int|WP_Error The post ID of the default fallback menu or a WP_Error object.
+	 */
+	public static function create_classic_menu_fallback() {
+		// See if we have a classic menu.
+		$classic_nav_menu = static::get_classic_menu_fallback();
+
+		if ( ! $classic_nav_menu ) {
+			return new WP_Error( 'no_classic_menus', 'No Classic Menus found.' );
+		}
+
+		// Todo: inject as dependency.
+		$menu_converter = new WP_Menu_Conversion_Gutenberg( $classic_nav_menu );
```

Comment:
Currently we're depending on this function directly. Would it be preferable to inject the dependency (perhaps in the constructor?) as an `Interface`? What would folks recommend?

Reply:
Would we ever need to swap this converter with something else? We also depend on WP_Query higher in the code, unlikely that will also need to be decoupled too.

Reply:
I guess you never know. I've not seen any type of IoC used in WP Core so I'd need to be pointed at an example if there's an establish pattern. Otherwise we can just depend on the class implicitly like we do here already.